crates/walrus-sdk/src/client/auto_tune.rs

```diff
@@ -0,0 +1,373 @@
+///    at the peak throughput.
+///
+use std::{
+    sync::{Arc, Mutex},
+    time::{Duration, Instant},
+};
+
+use tokio::sync::{
+    Semaphore,
+    mpsc::{UnboundedSender, unbounded_channel},
+};
+use walrus_core::SliverType;
+
+use crate::config::communication_config::DataInFlightAutoTuneConfig;
+
+/// Handle used by write tasks to report progress to the auto tuner.
+#[derive(Clone, Debug)]
+pub(crate) struct AutoTuneHandle {
+    inner: Arc<AutoTuneInner>,
+    semaphore: Arc<Semaphore>,
+    initial_permits: usize,
+    primary_weight: f64,
+    secondary_weight: f64,
+}
+
+impl AutoTuneHandle {
+    pub fn new(config: &DataInFlightAutoTuneConfig, allowed_permits: usize) -> Option<Self> {
+        if !config.enabled {
+            return None;
+        }
+
+        let allowed_permits = allowed_permits.max(1);
+
+        let (config_min, config_max) = config.permit_bounds();
+        let max_permits = config_max.min(allowed_permits).max(1);
+        let min_permits = config_min.min(max_permits).max(1);
+
+        if max_permits <= 1 {
+            tracing::warn!(
+                target: "walrus::auto_tune",
+                max_permits,
+                "auto tune disabled because at most one concurrent write is allowed"
+            );
+            return None;
+        }
+
+        if min_permits >= max_permits {
+            tracing::warn!(
+                target: "walrus::auto_tune",
+                min_permits,
+                max_permits,
+                "auto tune disabled due to invalid permit bounds"
+            );
+            return None;
+        }
+
+        let sample_target = config.sample_target();
+        let window_timeout = config.timeout();
+        let increase_factor = config.increase_factor();
+        let lock_factor = config.lock_factor();
+        let primary_weight = 1.0_f64;
+        let secondary_weight = config.secondary_sliver_weight().min(primary_weight);
+
+        let initial_permits = min_permits;
+        let state = AutoTuneState::new(
+            initial_permits,
+            min_permits,
+            max_permits,
+            sample_target,
+            window_timeout,
+            increase_factor,
+            lock_factor,
+        );
+        let semaphore = Arc::new(Semaphore::new(initial_permits));
+
+        Some(Self {
+            inner: Arc::new(AutoTuneInner::new(semaphore.clone(), state)),
+            semaphore,
+            initial_permits,
+            primary_weight,
+            secondary_weight,
+        })
+    }
+
+    pub fn semaphore(&self) -> Arc<Semaphore> {
+        self.semaphore.clone()
+    }
+
+    pub fn initial_permits(&self) -> usize {
+        self.initial_permits
+    }
+
+    pub fn record_success(&self, bytes: usize, sliver_type: SliverType, completed_at: Instant) {
+        let weight = self.weight_for(sliver_type);
+        let mut state = self.inner.state.lock().expect("auto tune mutex poisoned");
```

Comment:
Nit: Normally we use a formulation describing why the expect should succeed.

```suggestion
        let mut state = self.inner.state.lock().expect("mutex should not be poisoned");
```

Reply:
done